### ISSUE TITLE
Should it be easier to link linux VM binaries against the buildroot's uclibc?

### ISSUE THREAD

**Issue by ANogin** [SEL4PROJ/camkes-vm-linux#3]
Currently, the build setup for the in-VM binaries in all examples seems to be limited to creating binaries statically linked against the host glibc. There are a few issues with it:
- This is a bit wasteful as now the rootfs gets a bunch of copies of two different libcs
- getaddrinfo does not work (the glibc version requires dynamic loading of nss libraries to work).

**Comment by kent-mcleod** [comment]
It would be ideal if this was the default. It requires creating a CMake toolchain file that specifies the buildroot uclibc sysroot as the system library path.  It may not be too hard to set this up manually on a per project basis, but currently the vm-linux build scripts that we have don't support this.

**Comment by ANogin** [comment]
Right, this seems like non-trivial amount of work; probably more of a long-term RFE. I was thinking something almost Dockerfile-like that could grab a specific version of buildroot, with some specific packages, and then build more on top of it.

P.S. My other reason for filing this RFE was a weird linking issue causing the static binary to crash on startup, which I did not want to bring it up until I had more time to investigate - now I learned that I can work around that particular issue by adding the magical chant of `-Wl,--whole-archive -lpthread -Wl,--no-whole-archive` to my `-static` link command line... Still, working getaddrinfo would have been nice...

**Comment by ANogin** [comment]
Actually, I take the "non-trivial amount of work" back - this turned out to be quite straightforward.
1. Build buildroot yourself, enabling cpio creation (`BR2_TARGET_ROOTFS_CPIO=y` and `BR2_TARGET_ROOTFS_CPIO_NONE=y`)
    - Have not tried, but should be fairly easy to incorporate this step into cmake, very similarly to how `projects/vm-linux/CMakeLists.txt` already downloads and configures Linux kernel sources

2. Replace `-DCMAKE_TOOLCHAIN_FILE=${LINUX_32BIT_TOOLCHAIN}` in relevant `ExternalProject_Add` to point to builtroot's `output/host/usr/share/buildroot/toolchainfile.cmake`

3. Use buildroot's output/images/rootfs.cpio as a starting point that overlays are added to.

4. That's it - just a few small cosmetic steps left, along the lines of 
    - make sure `-std=c11` is added to the command line (there by default in the previous toolchain file, but not in buildroot's - although maybe I just missed an option to add it).
    - remove `target_link_libraries(camkesdevice -static-libgcc -static)` 
    - add "#define _GNU_SOURCE" in a few source files (or `#define _POSIX_C_SOURCE 200809L` / `#define _XOPEN_SOURCE 500`).

Now the 10+MB `default_buildroot_rootfs.cpio` is replaced with my own of less than 5MB (and that with OpenSSL libraries and some other stuff I needed to add), and everything seems to work! And my next step might very well be to create a VM with a bunch of python packages - it should just be an easy tweak of buildroot's .config to get the rootfs for that!

**Comment by kent-mcleod** [comment]
This is great! We've been waiting for a chance to add better buildroot integration into the rest of the CMake build system to be able to more easily manage the dynamic linking compatibility across the different sysroot environment configurations (uclibc/glibc).  It's nice to know that with an existing buildroot toolchain installed somewhere it becomes easy to build guest apps against it.
